v24.1.0-rc.1 release notes (#18524)
cockroachdb/docs: src/current/_includes/releases/v24.1/v24.1.0-rc.1.md

## v24.1.0-rc.1

Release Date: May 8, 2024

{% include releases/new-release-downloads-docker-image.md release=include.release %}

<h3 id="v24-1-0-rc-1-sql-language-changes">SQL language changes</h3>

> - Added a new session setting `optimizer_use_improved_multi_column_selectivity_estimate`, which if enabled, causes the [optimizer]({% link v24.1/cost-based-optimizer.md %}) to use an improved selectivity estimate for multi-column predicates. This setting will default to `true` on v24.2 and later, and `false` on earlier versions. [#123106][#123106]

Reviewer: Is there a good link for "session settings"? Is that different than a session variable? Is there a good link for multi-column predicates?

<h3 id="v24-1-0-rc-1-operational-changes">Operational changes</h3>

> - Added two new metrics: `range.snapshots.upreplication.rcvd-bytes` counts the number of Raft recovery snapshot bytes received, and `range.snapshots.upreplication.sent-bytes` counts the number of Raft recovery snapshot bytes sent. Also updated `range.snapshots.recovery.rcvd-bytes` and `range.snapshots.recovery.sent-bytes` to only include Raft snapshots. A new line was added to the [**Snapshot Data Received**]({% link v24.1/ui-replication-dashboard.md %}#snapshot-data-received) graph. [#123055][#123055]

Reviewer: There could be links to Metrics and Raft, wdyt?

<h3 id="v24-1-0-rc-1-db-console-changes">DB Console changes</h3>

> - Added a **Replication Lag** graph, which measures replication lag between primary and standby clusters for [physical cluster replication]({% link v24.1/physical-cluster-replication-overview.md %}), to the [**Physical Cluster Replication**]({% link v24.1/physical-cluster-replication-monitoring.md %}) dashboard. [#123285][#123285]

Reviewer: ```suggestion
- Added a **Replication Lag** graph to the [**Physical Cluster Replication**]({% link v24.1/physical-cluster-replication-monitoring.md %}) dashboard to measure replication lag between primary and standby clusters using [physical cluster replication]({% link v24.1/physical-cluster-replication-overview.md %}). [#123285][#123285]
```

<h3 id="v24-1-0-rc-1-bug-fixes">Bug fixes</h3>

> - The [**Tables**]({% link v24.1/ui-databases-page.md %}#tables-view) and [**Table details**]({% link v24.1/ui-databases-page.md %}#table-details) pages in the DB Console now display the correct **Table Stats Last Updated** value. [#122816][#122816]

Reviewer: ```suggestion
- Fixed a bug that caused the [**Tables**]({% link v24.1/ui-databases-page.md %}#tables-view) and [**Table Details**]({% link v24.1/ui-databases-page.md %}#table-details) pages in the DB Console to display an incorrect value for **Table Stats Last Updated**. [#122816][#122816]
```

> - Fixed a bug in the DB Console's [**Custom Chart**]({% link v24.1/ui-custom-chart-debug-page.md %}) tool where store-level metrics were not being displayed properly. Previously, if a store-level metric was selected to be displayed at the store-level on a multi-store node, only data for the 1st store ID associated with that node would be displayed.  This patch ensures that data is displayed for all stores present on a node. Additionally, it updates the behavior to show a single time series for each store, as opposed to aggregating (e.g., `SUM`) all stores across the node. This allows finer-grained observability into store-level metrics. [#122705][#122705]

Reviewer: ```suggestion
- Fixed a bug in the DB Console's [**Custom Chart**]({% link v24.1/ui-custom-chart-debug-page.md %}) tool where store-level metrics were displayed only for the first store ID associated with the node. Now a single time series is shown for each store, rather than an aggregated value for all of the node's stores. This allows finer-grained monitoring of store-level metrics. [#122705][#122705]
```

Author: Edited your suggestion to add `data is displayed for all stores present on a node` which was also a key change here.

> - Previously, CockroachDB could run into `attempting to append refresh spans after the tracked timestamp has moved forward` internal errors in some edge cases. This is now fixed. The bug had been present since v22.2. [#123136][#123136]

Reviewer: ```suggestion
- Fixed a bug introduced in v22.2 that could cause the internal error `attempting to append refresh spans after the tracked timestamp has moved forward` in some edge cases.  [#123136][#123136]
```
There is more info in the code PR, but I'm not sure how to slot it into the note to make it more specific.

> - Fixed a bug where `TYPEDESC SCHEMA CHANGE` jobs could end up retrying forever if the descriptor targeted by them was already dropped. [#123273][#123273]

Reviewer: ```suggestion
- Fixed a bug where a `TYPEDESC SCHEMA CHANGE` job could retry forever if the descriptor it targeted was already dropped. [#123273][#123273]
```

> - Previously, if the legacy schema changer was being used, the [`CREATE SEQUENCE`]({% link v24.1/create-sequence.md %}) command would incorrectly check if the user had the `CREATE` privilege on the parent database. This check is now removed, and the `CREATE SEQUENCE` command only needs `CREATE` privilege on the parent schema now. [#123289][#123289]

Reviewer: ```suggestion
- Fixed a bug where, if the legacy schema changer was enabled, the [`CREATE SEQUENCE`]({% link v24.1/create-sequence.md %}) command would incorrectly require the user to have the `CREATE` [privilege]({% link {{ page.version.version }}/security-reference/authorization.md %}#privileges) on the parent database rather than only on the parent schema.[#123289][#123289]
```

Author: Phew nice

> - A job will now log rather than fail if it reports an out-of bound progress fraction. [#122965][#122965]

Reviewer: ```suggestion
- Fixed a bug where a [job]({% link {{ page.version.version }}/show-jobs.md %}) would fail if it reported an out-of-bound progress fraction. The error is now logged no lo nger causes the job to fail. [#122965][#122965]
```

<h3 id="v24-1-0-rc-1-performance-improvements">Performance improvements</h3>

> - Added a new session setting `optimizer_use_improved_zigzag_join_costing`. When enabled, the cost of zigzag joins is updated so they will be never be chosen over scans unless they produce fewer rows. This change only matters if the setting `enable_zigzag_join` is also true. [#123106][#123106]

Reviewer: ```suggestion
- Added a new session setting `optimizer_use_improved_zigzag_join_costing`. When enabled and when the cluster setting `enable_zigzag_join` is also enabled, the cost of zigzag joins is updated, and a zigzag join will be used only if it produces fewer rows than a scan.[#123106][#123106]
```
Link to "session setting" (and to "cluster setting" if you also take the suggestion)?

> - Improved the selectivity estimation of multi-column filters when the multi-column distinct count is high. This avoids cases where CockroachDB significantly over-estimates the selectivity of a multi-column predicate, thus preventing the [optimizer]({% link v24.1/cost-based-optimizer.md %}) from choosing a bad query plan. [#123106][#123106]

Reviewer: Link to multi-column filter and multi-column predicate (if they are different)?
```suggestion
- Improved the selectivity estimation of multi-column filters when the multi-column distinct count is high. This prevents [optimizer]({% link v24.1/cost-based-optimizer.md %}) from choosing a bad query plan due to over-estimating the selectivity of a multi-column predicate. [#123106][#123106]
```

> - Made error handling in the [vectorized execution engine]({% link v24.1/vectorized-execution.md %}) much cheaper. This should help avoid bad metastable regimes where all CPU time is consumed by statement timeout handling, leading to more statement timeouts. [#123501][#123501]

Reviewer: ```suggestion
- Improved the efficiency of error handling in the [vectorized execution engine]({% link v24.1/vectorized-execution.md %}), to reduce the CPU overhead of statement timeout handling and reduce the potential for statement timeouts. [#123501][#123501]
```

> - Disabled a [changefeed]({% link v24.1/change-data-capture-overview.md %}) optimization on reducing duplicates during aggregator restarts due to its bad performance. [#123597][#123597]

Reviewer: ```suggestion
- Disabled a poorly-performing [changefeed]({% link v24.1/change-data-capture-overview.md %}) optimization that was intended to reduce duplicates during aggregator restarts. [#123597][#123597]
```

<div class="release-note-contributors" markdown="1">

<h3 id="v24-1-0-rc-1-contributors">Contributors</h3>

This release includes 57 merged PRs by 24 authors.

</div>

[#122705]: https://github.com/cockroachdb/cockroach/pull/122705
[#122816]: https://github.com/cockroachdb/cockroach/pull/122816
[#122965]: https://github.com/cockroachdb/cockroach/pull/122965
[#123055]: https://github.com/cockroachdb/cockroach/pull/123055
[#123106]: https://github.com/cockroachdb/cockroach/pull/123106
[#123136]: https://github.com/cockroachdb/cockroach/pull/123136
[#123144]: https://github.com/cockroachdb/cockroach/pull/123144
[#123273]: https://github.com/cockroachdb/cockroach/pull/123273
[#123285]: https://github.com/cockroachdb/cockroach/pull/123285
[#123289]: https://github.com/cockroachdb/cockroach/pull/123289
[#123373]: https://github.com/cockroachdb/cockroach/pull/123373
[#123501]: https://github.com/cockroachdb/cockroach/pull/123501
[#123597]: https://github.com/cockroachdb/cockroach/pull/123597

> [1958fa48e]: https://github.com/cockroachdb/cockroach/commit/1958fa48e
> [5c5e596ff]: https://github.com/cockroachdb/cockroach/commit/5c5e596ff

Reviewer: ```suggestion
```
